### ISSUE TITLE
Why does Tracer.SpanBuilder extend SpanContext?

### ISSUE THREAD

**Issue by yurishkuro** [opentracing/opentracing-java#55]
@michaelsembwever it was added in otherwise non-API change #25, but no documentation why this is necessary.

**Comment by mabn** [comment]
I'm implementing a tracer based on opentracing-impl and this is one of the confusing parts.

Because of this the signature of Extractor is:
```
public interface Extractor<C> {
    Tracer.SpanBuilder extract(C carrier);
}
```
which does not make sense until you remember that SpanBuilder implements SpanContext.

Additionally the intention is to be able to *extract* SpanContext and use it as a parent for new span:
```
SpanContext parent = extractor.extract(carrier);
spanBuilder.childOf(parent)
```
which in effect loses most of the state of the "parent" SpanBuilder implicitly.

There's also a thing with context identifiers (traceId, parentSpanId, spanId) - obviously SpanContext needs to hold identifiers, but SpanBuilder does not. There's `AbstractTracer.createSpanBuilder(String operationName)` method which does not allow to pass any identifiers, which is good, but for example *Extractor* implementations must supply identifiers.

IMO it should be a separate type - AbstractSpanContext and implementations should provide their specific subtype with the identifiers they need. It would make things more clear.

**Comment by yurishkuro** [comment]
@mabn I agree with you. Do you want to try this change in a pull request? It requires some refactoring of the abstract impl.

I am not sure if the AbstractSpanContext is actually needed or not, I guess the only functionality it can provide is the baggage storage, but without a concrete constructor it's difficult to implement something like `SpanContext withBaggage(k, v String)` method, that was used in implementations in other languages in order to keep the context immutable.

**Comment by bensigelman** [comment]
@mabn +1 that this is odd... kind of slipped past me in a PR but it seems odd to me conceptually and I don't understand why it was deemed necessary as part of #25.

**Comment by michaelsembwever** [comment]
> @mabn I agree with you. Do you want to try this change in a pull request? It requires some refactoring of the abstract impl.

If you could at the same time provide the PR against https://github.com/openzipkin/brave-opentracing that would be appreciated.

I think https://github.com/openzipkin/brave-opentracing/pull/5 needs to be re-addressed here as well.

**Comment by mabn** [comment]
I'll give it a shot

**Comment by michaelsembwever** [comment]
> I think openzipkin/brave-opentracing#5 needs to be re-addressed here as well.

PR#5 just seems to have been forgotten .I think I'll have it merged in the next few minutes… So  expect to base any changes off it.

**Comment by michaelsembwever** [comment]
> .I think I'll have it merged in the next few minutes… So expect to base any changes off it.

Will happen in a few hours. master has been updated though to use latest opentracing and brave.

**Comment by michaelsembwever** [comment]
https://github.com/openzipkin/brave-opentracing/pull/5 has been merged.

**Comment by bhs** [comment]
Closed via #121.
